### PR TITLE
feat: cleanup some Nat lemmas

### PR DESCRIPTION
**Pull request by fgdorais**
Depends on #182 

Potentially breaking changes:

* Changed `Nat.mul_lt_mul` (1 use in Mathlib). See comments below.
* Made two parameters implicit in `Nat.lt_add_right` (4 uses in Mathlib)
* Made parameters implicit for `Nat.add_le_add_iff_left` (no uses in Mathlib)
* Made parameters implicit for `Nat.add_le_add_iff_right` (no uses in Mathlib)
* Made parameters implicit for `Nat.add_lt_add_iff_left` (no uses in Mathlib)
* Made parameters implicit for `Nat.add_lt_add_iff_right` (no uses in Mathlib)
* Made parameters implicit for `Nat.dvd_iff_mod_eq_zero` (16 uses in Mathlib)
* Renamed `Nat.max_le` (no uses in Mathlib) to `Nat.max_le_iff` and added new `Nat.max_le` (Mathlib conformance)
* Renamed `Nat.le_min` (no uses in Mathlib)  to `Nat.le_min_iff`  and added new `Nat.le_min` (Mathlib conformance)
* Protected `Nat.zero_le` (unknown uses in Mathlib; at least one hiding clause)
* Protected `Nat.one_pos` (unknown uses in Mathlib)
* Protected `Nat.two_pos` (no uses in Mathlib)
* Deprecated `Nat.lt_min` (no uses in Mathlib)
* Deprecated `Nat.add_le_to_le_sub` (no uses in Mathlib)
* Deprecated `Nat.succ_add_eq_succ_add` (4 uses in Mathlib)
* Deprecated `Nat.mul_le_mul_of_nonneg_left` (4 uses in Mathlib)
* Deprecated `Nat.mul_le_mul_of_nonneg_right` (3 uses in Mathlib)
* Deprecated `Nat.mul_lt_mul'` (2 uses in Mathlib)
* Deprecated `Nat.zero_max` (1 use in Mathlib)
* Deprecated `Nat.zero_min` (no uses in Mathlib)
* Deprecated `Nat.max_zero` (no uses in Mathlib)
* Deprecated `Nat.min_zero` (no uses in Mathlib)

Deprecated lemmas all have recommended replacements.

### CODE REVIEW

File: Std/Data/Nat/Lemmas.lean [hunk 1]
```diff
@@ -99,63 +100,55 @@ theorem succ_le {n m : Nat} : succ n ≤ m ↔ n < m := .rfl
...
-theorem exists_eq_succ_of_ne_zero {n : Nat} (H : n ≠ 0) : ∃ k, n = succ k :=
-  ⟨_, (eq_zero_or_eq_succ_pred _).resolve_left H⟩
+theorem exists_eq_succ_of_ne_zero : ∀ {n : Nat}, n ≠ 0 → ∃ k, n = succ k
+  | _+1, _ => ⟨_, rfl⟩
 
 theorem succ_eq_one_add (n : Nat) : n.succ = 1 + n := by
   rw [Nat.succ_eq_add_one, Nat.add_comm]
 
-theorem succ_inj' : succ n = succ m ↔ n = m :=
+theorem succ_inj' {n m : Nat} : succ n = succ m ↔ n = m :=
```
Comment (digama0) [inline Std/Data/Nat/Lemmas.lean:121]: What is the reason for this change? Style-wise, std prefers auto-implicits, although it's fine to include types if you think it is clearer or you need to manipulate argument order. I would prefer to leave them off in most cases here, it's pretty obvious everything is about Nat.

Comment (fgdorais) [reply Std/Data/Nat/Lemmas.lean]: Easy fix!

Comment (fgdorais) [reply Std/Data/Nat/Lemmas.lean]: Should I do the reverse and remove implicit parameters that could be auto-implicit?

Comment (digama0) [reply Std/Data/Nat/Lemmas.lean]: That's fine by me. Sometimes it makes a small difference with order of parameters, but I think we want to normalize to the ordering you get with autoimplicits anyway.

**Comment by fgdorais** [comment]
This is a good stopping point. I did only a few small additions. The main one are some custom recursors in Std.Data.Nat.Basic. I'm happy to hear reviews from anyone who has time.

**Comment by fgdorais** [comment]
Regarding #118: I'm tempted to rename `Nat.mul_lt_mul₀` to just `Nat.mul_lt_mul`. There are some issues:

- On the one hand, Mathlib has `mul_lt_mul₀` in `Mathlib.Algebra.Order.WithZero` so it makes sense to use the same name. On the other hand, like every theorem in that file, the subscript zero indicates `WithZero` which doesn't make much sense as a distinction in `Nat`.
- On the one hand, there is a deprecated `Nat.mul_lt_mul` with one use in Mathlib. But the old `Nat.mul_lt_mul` was a bad theorem with unnecessarily strong hypotheses, so it's a good riddance.

**Comment by fgdorais** [comment]
I finally opted for `Nat.mul_lt_mul` over `Nat.mul_lt_mul₀`. What tipped the scale for me is that the old `Nat.mul_lt_mul` was just a bad theorem. Existing uses can be replaced by `Nat.mul_lt_mul_of_lt_of_le` after adapting the hypotheses.

**Comment by fgdorais** [comment]
Now broken up into smaller pieces.
